### DISCUSSION TITLE
Add metadata to ROS2 MCAP file with rosbag2_py

### DISCUSSION THREAD

**Discussion by christian-nils** [foxglove/mcap#1233] [General] [answered]
Hello,

I would like to add some metadata to my MCAP file. As suggested in this issue: https://github.com/foxglove/mcap/issues/861, I am using rosbag2_py to write the MCAP file. I would like to add a couple of metadata to the file but I seem not to find the right way to do it using rosbag2_py only (but it seems to be possible using the MCAP Python library: https://mcap.dev/docs/python/_modules/mcap/writer#Writer.add_metadata). Anyone would know how I could do this? 

Thanks,

Christian-Nils

**Comment by jtbandes** [comment, marked as answer]
Hi there,
I believe it's correct that rosbag2 does not expose any way to work with MCAP metadata.

One possible solution is to use the MCAP CLI (`mcap add metadata`):

```
$ mcap help add
Add records to an existing MCAP file

Usage:
  mcap add [flags]
  mcap add [command]

Available Commands:
  attachment  Add an attachment to an MCAP file
  metadata    Add metadata to an MCAP file

Flags:
  -h, --help   help for add

Global Flags:
      --config string          Config file (default is $HOME/.mcap.yaml)
      --pprof-profile          Record pprof profiles of command execution. Profiles will be written to files: mcap-mem.prof, mcap-cpu.prof, and mcap-block.pprof. Defaults to false.
      --strict-message-order   Require that messages have a monotonic log time
  -v, --verbose                Verbose output

Use "mcap add [command] --help" for more information about a command.
```

**Reply by christian-nils** [reply]
Hi Jacob!

Thanks for your quick reply! I was trying to avoid passing by the CLI and keep using only the Python API. I tried to combine both rosbag2_py and MCAP Python library but it is not an option as I read somewhere that only the golang and typescript implementations allow appending MCAP files.

I also tried to only use the Python library when creating my MCAP file from scratch but it is not that straightforward to get the message definitions from the message Python class (I could not find the equivalent to ROS1 with `message.__class__.__full_text`)

I will go with using the mcap CLI then, it seems to be the cleanest option.

Thanks again for the reply, have a great day.
